Kinesis: The security token included in the request is expired

## Description

<summary>
When using Assume Role ARN functionality, the following error occurs after about one hour of the input running. Restarting the input fixes the issue.

Assume Role ARN support was added in release 3.1.4, but this bug exists, which makes it not function correctly.

The security token included in the request is expired
<details> 
2020-02-03 12:07:06,033 ERROR: software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseCoordinator - LeasingException encountered in lease taking thread
software.amazon.kinesis.leases.exceptions.DependencyException: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: The security token included in the request is expired (Service: DynamoDb, Status Code: 400, Request ID:)
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRefresher.list(DynamoDBLeaseRefresher.java:330) ~[graylog-plugin-integrations-3.2.0.jar:?]
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRefresher.listLeases(DynamoDBLeaseRefresher.java:260) ~[graylog-plugin-integrations-3.2.0.jar:?]
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseTaker.updateAllLeases(DynamoDBLeaseTaker.java:253) ~[graylog-plugin-integrations-3.2.0.jar:?]
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseTaker.takeLeases(DynamoDBLeaseTaker.java:151) ~[graylog-plugin-integrations-3.2.0.jar:?]
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseTaker.takeLeases(DynamoDBLeaseTaker.java:119) ~[graylog-plugin-integrations-3.2.0.jar:?]
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseCoordinator.runLeaseTaker(DynamoDBLeaseCoordinator.java:254) ~[graylog-plugin-integrations-3.2.0.jar:?]
        at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseCoordinator$TakerRunnable.run(DynamoDBLeaseCoordinator.java:185) [graylog-plugin-integrations-3.2.0.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_201]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:1.8.0_201]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_201]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:1.8.0_201]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_201]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_201]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201]
Caused by: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: The security token included in the request is expired (Service: DynamoDb, Status Code: 400, Request ID: CTIBM25HOFHUJ3COJ3GTE48LK7VV4KQNSO5AEMVJF66Q9ASUAAJG)
        at software.amazon.awssdk.services.dynamodb.model.DynamoDbException$BuilderImpl.build(DynamoDbException.java:95) ~[?:?]
        at software.amazon.awssdk.services.dynamodb.model.DynamoDbException$BuilderImpl.build(DynamoDbException.java:55) ~[?:?]
        at software.amazon.awssdk.protocols.json.internal.unmarshall.AwsJsonProtocolErrorUnmarshaller.unmarshall(AwsJsonProtocolErrorUnmarshaller.java:88) ~[?:?]
        at software.amazon.awssdk.protocols.json.internal.unmarshall.AwsJsonProtocolErrorUnmarshaller.handle(AwsJsonProtocolErrorUnmarshaller.java:63) ~[?:?]
        at software.amazon.awssdk.protocols.json.internal.unmarshall.AwsJsonProtocolErrorUnmarshaller.handle(AwsJsonProtocolErrorUnmarshaller.java:42) ~[?:?]
        at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:88) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952) ~[?:1.8.0_201]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926) ~[?:1.8.0_201]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:1.8.0_201]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962) ~[?:1.8.0_201]
        at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:129) ~[?:?]
        at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:171) ~[?:?]
        at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:68) ~[?:?]
        at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:287) ~[?:?]
        at software.amazon.awssdk.thirdparty.com.typesafe.netty.HandlerPublisher.publishMessage(HandlerPublisher.java:362) ~[?:?]
        at software.amazon.awssdk.thirdparty.com.typesafe.netty.HandlerPublisher.flushBuffer(HandlerPublisher.java:304) ~[?:?]
        at software.amazon.awssdk.thirdparty.com.typesafe.netty.HandlerPublisher.receivedDemand(HandlerPublisher.java:258) ~[?:?]
        at software.amazon.awssdk.thirdparty.com.typesafe.netty.HandlerPublisher.access$200(HandlerPublisher.java:41) ~[?:?]
        at software.amazon.awssdk.thirdparty.com.typesafe.netty.HandlerPublisher$ChannelSubscription$1.run(HandlerPublisher.java:452) ~[?:?]
        at software.amazon.awssdk.thirdparty.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[?:?]
        at software.amazon.awssdk.thirdparty.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510) ~[?:?]
        at software.amazon.awssdk.thirdparty.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:518) ~[?:?]
        at software.amazon.awssdk.thirdparty.io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044) ~[?:?]
        at software.amazon.awssdk.thirdparty.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[?:?]
        ... 1 more
</details>
</summary>

## Steps To Reproduce

Use the Assume Role ARN for the new AWS Kinesis/Cloudwatch integration.

## Environment

- Graylog Version:
- Elasticsearch Version:
- MongoDB Version:
- Browser Version:

--- a/src/main/java/org/graylog/integrations/aws/AWSAuthFactory.java
+++ b/src/main/java/org/graylog/integrations/aws/AWSAuthFactory.java
@@ -31,9 +31,9 @@ public class AWSAuthFactory {
      * Javadoc for more information.
      */
     public static AwsCredentialsProvider getAuthProvider(@Nullable String accessKey,
-                                                  @Nullable String secretKey,
-                                                  @Nullable String stsRegion,
-                                                  @Nullable String assumeRoleArn) {
+                                                         @Nullable String secretKey,
+                                                         @Nullable String stsRegion,
+                                                         @Nullable String assumeRoleArn) {
         AwsCredentialsProvider awsCredentials;
         if (!isNullOrEmpty(accessKey) && !isNullOrEmpty(secretKey)) {
             LOG.debug("Using explicitly provided key and secret.");
@@ -59,7 +59,7 @@ public class AWSAuthFactory {
      * permission, which provides authorization for a role to be assumed.
      */
     private static AwsCredentialsProvider buildStsCredentialsProvider(AwsCredentialsProvider awsCredentials, String stsRegion,
-                                                               String assumeRoleArn, @Nullable String accessKey) {
+                                                                      String assumeRoleArn, @Nullable String accessKey) {
 
         StsClient stsClient = StsClient.builder().region(Region.of(stsRegion)).credentialsProvider(awsCredentials).build();
 

--- a/src/main/java/org/graylog/integrations/aws/AWSAuthFactory.java
+++ b/src/main/java/org/graylog/integrations/aws/AWSAuthFactory.java
@@ -3,7 +3,6 @@ package org.graylog.integrations.aws;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
-import software.amazon.awssdk.auth.credentials.AwsCredentials;
 import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
 import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
 import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
@@ -17,27 +16,24 @@ import javax.annotation.Nullable;
 
 import static com.google.common.base.Strings.isNullOrEmpty;
 
-/**
- * Resolves the appropriate AWS authorization provider.
- *
- * If an {@code accessKey} and {@code secretKey} are provided, they will be used explicitly.
- * If not, the default DefaultCredentialsProvider will be used instead. This will resolve the role/policy
- * using Java props, environment variables, EC2 instance roles etc. See the {@link DefaultCredentialsProvider}
- * Javadoc for more information.
- */
-public class AWSAuthProvider implements AwsCredentialsProvider {
-    private static final Logger LOG = LoggerFactory.getLogger(AWSAuthProvider.class);
+public class AWSAuthFactory {
+    private static final Logger LOG = LoggerFactory.getLogger(AWSAuthFactory.class);
 
-    private AwsCredentials credentials;
-
-    public AWSAuthProvider(@Nullable String stsRegion, @Nullable String accessKey, @Nullable String secretKey, @Nullable String assumeRoleArn) {
-        this.credentials = this.resolveAuthentication(accessKey, secretKey, stsRegion, assumeRoleArn);
+    private AWSAuthFactory() {
     }
 
-    private AwsCredentials resolveAuthentication(@Nullable String accessKey,
-                                                 @Nullable String secretKey,
-                                                 @Nullable String stsRegion,
-                                                 @Nullable String assumeRoleArn) {
+    /**
+     * Resolves the appropriate AWS authorization provider based on the input.
+     *
+     * If an {@code accessKey} and {@code secretKey} are provided, they will be used explicitly.
+     * If not, the default DefaultCredentialsProvider will be used instead. This will resolve the role/policy
+     * using Java props, environment variables, EC2 instance roles etc. See the {@link DefaultCredentialsProvider}
+     * Javadoc for more information.
+     */
+    public static AwsCredentialsProvider getAuthProvider(@Nullable String accessKey,
+                                                  @Nullable String secretKey,
+                                                  @Nullable String stsRegion,
+                                                  @Nullable String assumeRoleArn) {
         AwsCredentialsProvider awsCredentials;
         if (!isNullOrEmpty(accessKey) && !isNullOrEmpty(secretKey)) {
             LOG.debug("Using explicitly provided key and secret.");
@@ -50,11 +46,10 @@ public class AWSAuthProvider implements AwsCredentialsProvider {
         // Apply the Assume Role ARN Authorization if specified. All AWSCredentialsProviders support this.
         if (!isNullOrEmpty(assumeRoleArn) && !isNullOrEmpty(stsRegion)) {
             LOG.debug("Creating cross account assume role credentials");
-            return buildStsCredentialsProvider(awsCredentials, stsRegion, assumeRoleArn, accessKey)
-                    .resolveCredentials();
+            return buildStsCredentialsProvider(awsCredentials, stsRegion, assumeRoleArn, accessKey);
         }
 
-        return awsCredentials.resolveCredentials();
+        return awsCredentials;
     }
 
     /**
@@ -63,7 +58,7 @@ public class AWSAuthProvider implements AwsCredentialsProvider {
      * Note: In order to assume a role, a role must be provided to the AWS STS client a role that has the "sts:AssumeRole"
      * permission, which provides authorization for a role to be assumed.
      */
-    private AwsCredentialsProvider buildStsCredentialsProvider(AwsCredentialsProvider awsCredentials, String stsRegion,
+    private static AwsCredentialsProvider buildStsCredentialsProvider(AwsCredentialsProvider awsCredentials, String stsRegion,
                                                                String assumeRoleArn, @Nullable String accessKey) {
 
         StsClient stsClient = StsClient.builder().region(Region.of(stsRegion)).credentialsProvider(awsCredentials).build();
@@ -86,10 +81,5 @@ public class AWSAuthProvider implements AwsCredentialsProvider {
                                                                                           .build())
                                                .stsClient(stsClient)
                                                .build();
-    }
-
-    @Override
-    public AwsCredentials resolveCredentials() {
-        return credentials;
     }
 }

--- a/src/main/java/org/graylog/integrations/aws/AWSAuthFactory.java
+++ b/src/main/java/org/graylog/integrations/aws/AWSAuthFactory.java
@@ -19,9 +19,6 @@ import static com.google.common.base.Strings.isNullOrEmpty;
 public class AWSAuthFactory {
     private static final Logger LOG = LoggerFactory.getLogger(AWSAuthFactory.class);
 
-    private AWSAuthFactory() {
-    }
-
     /**
      * Resolves the appropriate AWS authorization provider based on the input.
      *
@@ -30,10 +27,10 @@ public class AWSAuthFactory {
      * using Java props, environment variables, EC2 instance roles etc. See the {@link DefaultCredentialsProvider}
      * Javadoc for more information.
      */
-    public static AwsCredentialsProvider getAuthProvider(@Nullable String accessKey,
-                                                         @Nullable String secretKey,
-                                                         @Nullable String stsRegion,
-                                                         @Nullable String assumeRoleArn) {
+    public static AwsCredentialsProvider create(@Nullable String accessKey,
+                                                @Nullable String secretKey,
+                                                @Nullable String stsRegion,
+                                                @Nullable String assumeRoleArn) {
         AwsCredentialsProvider awsCredentials;
         if (!isNullOrEmpty(accessKey) && !isNullOrEmpty(secretKey)) {
             LOG.debug("Using explicitly provided key and secret.");

--- a/src/main/java/org/graylog/integrations/aws/AWSClientBuilderUtil.java
+++ b/src/main/java/org/graylog/integrations/aws/AWSClientBuilderUtil.java
@@ -56,8 +56,8 @@ public class AWSClientBuilderUtil {
         AWSClientBuilderUtil.initializeBuilder(clientBuilder,
                                                request.cloudwatchEndpoint(),
                                                Region.of(request.region()),
-                                               new AWSAuthProvider(request.region(), request.awsAccessKeyId(),
-                                                                   request.awsSecretAccessKey(), request.assumeRoleArn()));
+                                               AWSAuthFactory.getAuthProvider(request.region(), request.awsAccessKeyId(),
+                                                                  request.awsSecretAccessKey(), request.assumeRoleArn()));
 
         return clientBuilder.build();
     }
@@ -74,8 +74,8 @@ public class AWSClientBuilderUtil {
         AWSClientBuilderUtil.initializeBuilder(clientBuilder,
                                                request.kinesisEndpoint(),
                                                Region.of(request.region()),
-                                               new AWSAuthProvider(request.region(), request.awsAccessKeyId(),
-                                                                   request.awsSecretAccessKey(), request.assumeRoleArn()));
+                                               AWSAuthFactory.getAuthProvider(request.region(), request.awsAccessKeyId(),
+                                                                  request.awsSecretAccessKey(), request.assumeRoleArn()));
 
         return clientBuilder.build();
     }
@@ -92,9 +92,9 @@ public class AWSClientBuilderUtil {
         AWSClientBuilderUtil.initializeBuilder(clientBuilder,
                                                request.iamEndpoint(),
                                                Region.AWS_GLOBAL, // Always specify the global region for the IAM client.
-                                               new AWSAuthProvider(request.region(), // The AWSAuthProvider must still use the user-specified region, since a role might need to be assumed in that region.
-                                                                   request.awsAccessKeyId(),
-                                                                   request.awsSecretAccessKey(), request.assumeRoleArn()));
+                                               AWSAuthFactory.getAuthProvider(request.region(), // The AWSAuthProvider must still use the user-specified region, since a role might need to be assumed in that region.
+                                                                  request.awsAccessKeyId(),
+                                                                  request.awsSecretAccessKey(), request.assumeRoleArn()));
         return clientBuilder.build();
     }
 }

--- a/src/main/java/org/graylog/integrations/aws/AWSClientBuilderUtil.java
+++ b/src/main/java/org/graylog/integrations/aws/AWSClientBuilderUtil.java
@@ -56,8 +56,8 @@ public class AWSClientBuilderUtil {
         AWSClientBuilderUtil.initializeBuilder(clientBuilder,
                                                request.cloudwatchEndpoint(),
                                                Region.of(request.region()),
-                                               AWSAuthFactory.getAuthProvider(request.region(), request.awsAccessKeyId(),
-                                                                  request.awsSecretAccessKey(), request.assumeRoleArn()));
+                                               AWSAuthFactory.create(request.region(), request.awsAccessKeyId(),
+                                                                     request.awsSecretAccessKey(), request.assumeRoleArn()));
 
         return clientBuilder.build();
     }
@@ -74,8 +74,8 @@ public class AWSClientBuilderUtil {
         AWSClientBuilderUtil.initializeBuilder(clientBuilder,
                                                request.kinesisEndpoint(),
                                                Region.of(request.region()),
-                                               AWSAuthFactory.getAuthProvider(request.region(), request.awsAccessKeyId(),
-                                                                  request.awsSecretAccessKey(), request.assumeRoleArn()));
+                                               AWSAuthFactory.create(request.region(), request.awsAccessKeyId(),
+                                                                     request.awsSecretAccessKey(), request.assumeRoleArn()));
 
         return clientBuilder.build();
     }
@@ -92,9 +92,9 @@ public class AWSClientBuilderUtil {
         AWSClientBuilderUtil.initializeBuilder(clientBuilder,
                                                request.iamEndpoint(),
                                                Region.AWS_GLOBAL, // Always specify the global region for the IAM client.
-                                               AWSAuthFactory.getAuthProvider(request.region(), // The AWSAuthProvider must still use the user-specified region, since a role might need to be assumed in that region.
-                                                                  request.awsAccessKeyId(),
-                                                                  request.awsSecretAccessKey(), request.assumeRoleArn()));
+                                               AWSAuthFactory.create(request.region(), // The AWSAuthProvider must still use the user-specified region, since a role might need to be assumed in that region.
+                                                                     request.awsAccessKeyId(),
+                                                                     request.awsSecretAccessKey(), request.assumeRoleArn()));
         return clientBuilder.build();
     }
 }

--- a/src/main/java/org/graylog/integrations/aws/transports/KinesisConsumer.java
+++ b/src/main/java/org/graylog/integrations/aws/transports/KinesisConsumer.java
@@ -4,7 +4,7 @@ package org.graylog.integrations.aws.transports;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.base.Preconditions;
 import org.apache.commons.lang3.StringUtils;
-import org.graylog.integrations.aws.AWSAuthProvider;
+import org.graylog.integrations.aws.AWSAuthFactory;
 import org.graylog.integrations.aws.AWSClientBuilderUtil;
 import org.graylog.integrations.aws.AWSMessageType;
 import org.graylog.integrations.aws.resources.requests.AWSRequest;
@@ -76,8 +76,8 @@ public class KinesisConsumer implements Runnable {
     public void run() {
 
         LOG.debug("Starting the Kinesis Consumer.");
-        AwsCredentialsProvider credentialsProvider = new AWSAuthProvider(request.region(), request.awsAccessKeyId(),
-                                                                         request.awsSecretAccessKey(), request.assumeRoleArn());
+        AwsCredentialsProvider credentialsProvider = AWSAuthFactory.getAuthProvider(request.region(), request.awsAccessKeyId(),
+                                                                                    request.awsSecretAccessKey(), request.assumeRoleArn());
 
         final Region region = Region.of(request.region());
 

--- a/src/main/java/org/graylog/integrations/aws/transports/KinesisConsumer.java
+++ b/src/main/java/org/graylog/integrations/aws/transports/KinesisConsumer.java
@@ -76,8 +76,8 @@ public class KinesisConsumer implements Runnable {
     public void run() {
 
         LOG.debug("Starting the Kinesis Consumer.");
-        AwsCredentialsProvider credentialsProvider = AWSAuthFactory.getAuthProvider(request.region(), request.awsAccessKeyId(),
-                                                                                    request.awsSecretAccessKey(), request.assumeRoleArn());
+        AwsCredentialsProvider credentialsProvider = AWSAuthFactory.create(request.region(), request.awsAccessKeyId(),
+                                                                           request.awsSecretAccessKey(), request.assumeRoleArn());
 
         final Region region = Region.of(request.region());
 

--- a/src/test/java/org/graylog/integrations/aws/AWSAuthFactoryTest.java
+++ b/src/test/java/org/graylog/integrations/aws/AWSAuthFactoryTest.java
@@ -1,0 +1,21 @@
+package org.graylog.integrations.aws;
+
+import org.junit.Assert;
+import org.junit.Test;
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+
+public class AWSAuthFactoryTest {
+
+    @Test
+    public void testAutomaticAuth() {
+
+        Assert.assertTrue(AWSAuthFactory.create(null, null, null, null) instanceof DefaultCredentialsProvider);
+    }
+
+    @Test
+    public void testKeySecret() {
+
+        Assert.assertTrue(AWSAuthFactory.create("key", "secret", null, null) instanceof StaticCredentialsProvider);
+    }
+}

